ddtrace/tracer/tracer.go

```diff
@@ -156,16 +156,24 @@ func Inject(ctx ddtrace.SpanContext, carrier interface{}) error {
 
 // SetUser associates user information to the current trace which the
 // provided span belongs to. The options can be used to tune which user
-// bit of information gets monitored.
+// bit of information gets monitored. In case of distributed traces,
+// the user id can be propagated across traces using the WithPropagation() option.
+// See https://docs.datadoghq.com/security_platform/application_security/setup_and_configure/?tab=set_user#add-user-information-to-traces
 func SetUser(s Span, id string, opts ...UserMonitoringOption) {
 	if s == nil {
 		return
 	}
-	if span, ok := s.(*span); ok && span.context != nil {
-		span = span.context.trace.root
-		s = span
-	}
-	s.SetTag("usr.id", id)
+	if sp, ok := s.(*span); ok && sp.context != nil {
+		sp = sp.context.trace.root
+		// Unset the propagated user ID by default so that if the function is called without WithPropagation(),
+		// a propagated user ID coming from upstream won't be propagated anymore.
+		sp.Lock()
+		delete(sp.context.trace.propagatingTags, keyPropagatedUserID)
+		delete(sp.Meta, keyPropagatedUserID)
+		sp.Unlock()
+		s = sp
```

Comment:
I would make a helper function for this in `sp.context`. It's very unusual to handle locks in options.

Reply:
Just to circle back here: I think that synchronisation logic should be a low level thing and that it should not be acceptable to access it at such a high level, such as a functional API option.

It is best to think of the API in layers:
* User API: the stuff that the user calls directly. This should ideally be nothing but structs.
* Tracer API: the API that is used to interact with the core tracer, the one which integrations or external components would call.
* Core API: the (most likely unexported) internal API of the tracer.

I think that synchronisation mechanisms relating to the "Core API" being called in the "User API" is a very unhealthy decoupling and a smell, aka asking for trouble.

Just my 2c.

Reply:
After our slack conversation, here's what's changed:
- user monitoring options now set a `UserMonitoringConfig` struct instead of modifying spans on the fly
- `SetUser()` now builds the configuration and calls down to `span.setUser()`
- `span.setUser()` locks the span and sets/unsets the tags according to the configuration